osquery/tables/system/darwin/cups_jobs.cpp

```diff
@@ -0,0 +1,53 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <cups/cups.h>
+
+#include <osquery/system.h>
+#include <osquery/tables.h>
+
+namespace osquery {
+namespace tables {
+
+class SafeCupsJobs {
```

Comment:
I feel like this and the `SafeCupsDestination` class above are trying to accomplish what smart pointers could do for you? Considering the only logic you seem to be taking advantage of here with the class is achieving RIAA, maybe classes aren't needed?

Reply:
Is there a convenient way to have smart pointers call the specialized `cupsFreeJobs` function?

Reply:
@zwass You can create a lambda function and have that called when the smart pointer goes out of scope but you also need the number of elements in the array (because it doesn't store it itself). 

I originally solved this by have a structure that contained both and then have the smart pointer call a teardown when it stopped existing. The issue with this was that I now had a structure and a tear down function that conceptually related to each other but wasn't shown in the code. I then moved to a class so that function could be encapsulated as the destructor.

Reply:
I like it :) Thanks for the discussion.

Reply:
I have a new change coming in a few minutes which should address this concern along with decltype in one abstracted swoop :)